firedrake/output.py

```diff
@@ -651,3 +756,366 @@ def write(self, *functions, **kwargs):
+            self.save_function(_f)
+            return
+        vec = tf.dat._vec
+        vec.setName(tf.name())
+        dm.globalVectorView(self.viewer, sectiondm, vec)
+
+    def load_mesh(self, name=DEFAULT_MESH_NAME, distribution_parameters={}):
+        dist_key = hash(frozenset(distribution_parameters.items()))
+        mesh_key = (name, dist_key)
+        if mesh_key in self._meshes:
+            return self._meshes[mesh_key]["mesh"]
+        plex, sfXB = _dmload_plex(self.viewer, name)
+        # These labels are distribution dependent.
+        # We should be able to save/load labels selectively.
+        plex.removeLabel("pyop2_core")
+        plex.removeLabel("pyop2_owned")
+        plex.removeLabel("pyop2_ghost")
+        mesh = Mesh(plex, name=name, distribution_parameters=distribution_parameters)
+        self._meshes[mesh_key] = {"mesh": mesh,
+                                  "sfXB": sfXB}
+        return mesh
+
+    def load_function_space(self, name, mesh_name, distribution_parameters={}):
+        # Deal with `MixedFunctionSpace`s
+        if PREFIX + "_mixed_function_spaces" in self.h5pyfile["/topologies"]:
+            base_base_path = os.path.join("/topologies",
+                                          PREFIX + "_mixed_function_spaces")
+            if name in self.h5pyfile[base_base_path]:
+                base_path = os.path.join(base_base_path, name)
+                n = self.get_attr(base_path, PREFIX + "_num_sub_spaces")
+                Vsub_list = []
+                for i in range(n):
+                    path = os.path.join(base_path, str(i))
+                    Vsub_mesh_name = self.get_attr(path, PREFIX + "_mesh")
+                    Vsub_name = self.get_attr(path, PREFIX + "_function_space")
+                    Vsub = self.load_function_space(Vsub_name,
+                                                    Vsub_mesh_name,
+                                                    distribution_parameters=distribution_parameters)
+                    Vsub_list.append(Vsub)
+                V = functools.reduce(lambda a, b: a * b, Vsub_list)
+                V.name = name
+                return V
+        dist_key = hash(frozenset(distribution_parameters.items()))
+        mesh_key = (mesh_name, dist_key)
+        fs_key = (name, ) + mesh_key
+        if fs_key in self._function_spaces:
+            return self._function_spaces[fs_key]
+        # Load mesh
+        mesh = self.load_mesh(mesh_name, distribution_parameters=distribution_parameters)
+        mesh.init()
+        # Load element
+        sectiondm_name = self.find_sectiondm_name(mesh_name, name)
+        path = os.path.join("/topologies",
+                            mesh.topology_dm.getName(),
+                            "dms",
+                            sectiondm_name,
+                            PREFIX + "_function_spaces",
+                            name)
+        element_b = self.get_attr(path, PREFIX + "_ufl_element").tobytes()
```

Comment:
There's kind of a lot of repeated code for name construction in the hierarchy. I wonder if there is a way to make this less fragile (not specifically this line, I just needed to put this comment somewhere).

Reply:
I think it looks better now.